src/commands/run.rs

```diff
@@ -784,19 +784,24 @@ impl RunCommand {
             builder.allow_udp(enable);
         }
 
-        store.data_mut().preview2_ctx = Some(Arc::new(builder.build()));
+        let ctx = builder.build();
+        store.data_mut().preview2_ctx = Some(Arc::new(Mutex::new(ctx)));
         Ok(())
     }
 }
 
 #[derive(Default, Clone)]
 struct Host {
     preview1_ctx: Option<wasmtime_wasi::WasiCtx>,
-    preview2_ctx: Option<Arc<preview2::WasiCtx>>,
+
+    // The Mutex is only needed to satisfy the Sync constraint but we never
+    // actually perform any locking on it as we use Mutex::get_mut for every
+    // access.
+    preview2_ctx: Option<Arc<Mutex<preview2::WasiCtx>>>,
 
     // Resource table for preview2 if the `preview2_ctx` is in use, otherwise
     // "just" an empty table.
-    preview2_table: Arc<wasmtime::component::ResourceTable>,
+    preview2_table: Arc<Mutex<wasmtime::component::ResourceTable>>,
```

Comment:
Needing a mutex here is a minor bummer, but I think the next refactor is to make there be two different ctx structs:`Host` for all preview 2 compatible things, and `LegacyHost` for using wasi-common if, and only if, threads are required.

I am working on a PR that separates out wasi-common from wasmtime-wasi, so that we can ship just the preview 2 implementation without any of the old one, so I'll fix this up after that lands.

Reply:
Yeah, but at least `.lock()` never gets called, so effectively it's mostly a Box